### PR TITLE
Upstream config memory optimization

### PR DESCRIPTION
**Pull request by DiazAlan**
Commit Message: Reduce memory usage if cluster does not contain an upstream config
Additional Description: In cases where a cluster does not have a set upstream config, this will save memory using unique pointers
Risk Level: Low
Testing: Changed mocks as needed
Docs Changes: NA
Release Notes: NA
Platform Specific Features: NA

### CODE REVIEW

**Comment by DiazAlan** [comment]
@yurykats @adisuissa

File: source/common/upstream/upstream_impl.h
Comment (jmarantz) [inline source/common/upstream/upstream_impl.h:803]: nit: I feel like we should be returning an `OptRef<const envoy::config::core::v3::TypedExtensionConfig>` given that this is a const method.

Would that work at call-sites?

The version checked in now actually returned a const optional ref, so I think this PR should also return a config that cannot be modified by the caller.

**Review by jmarantz — COMMENTED** [review]
thanks for doing this; I have a drive-by comment.

File: envoy/upstream/upstream.h
Comment (adisuissa) [inline envoy/upstream/upstream.h:1010]: nit: return const TypedExtensionConfig

File: source/common/upstream/upstream_impl.h
Comment (adisuissa) [reply source/common/upstream/upstream_impl.h:803]: +1

File: source/common/upstream/upstream_impl.h
Comment (adisuissa) [inline source/common/upstream/upstream_impl.h:962]: nit: const

**Review by adisuissa — COMMENTED** [review]
Thanks! LGTM modulo using const.
+1 on @jmarantz comment, should prefer const if possible.

/wait

File: envoy/upstream/upstream.h
Comment (DiazAlan) [reply envoy/upstream/upstream.h]: Done

File: source/common/upstream/upstream_impl.h
Comment (adisuissa) [reply source/common/upstream/upstream_impl.h:962]: Can this also be `const TypedExtensionConfig`?

**Review by adisuissa — COMMENTED** [review]
Thanks!
One more nit that needs to be addressed.

File: source/common/upstream/upstream_impl.h
Comment (adisuissa) [reply source/common/upstream/upstream_impl.h:962]: After discussing this IRL it seems that this is non-const because of the tests, and I think that keeping it non-const is ok for now.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
cc @jmarantz if there are additional comments.

/assign-from @envoyproxy/senior-maintainers

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
